contracts/templates/ERC20Template3.sol

```diff
@@ -0,0 +1,1184 @@
+     */
+    function getFixedRates() public view returns (fixedRate[] memory) {
+        return (fixedRateExchanges);
+    }
+
+    /**
+     * @dev getDispensers
+     *      Returns the list of dispensers created for this datatoken
+    */
+    function getDispensers() public view returns (address[] memory) {
+        return (dispensers);
+    }
+
+    function _pullUnderlying(
+        address erc20,
+        address from,
+        address to,
+        uint256 amount
+    ) internal {
+        uint256 balanceBefore = IERC20(erc20).balanceOf(to);
+        IERC20(erc20).safeTransferFrom(from, to, amount);
+        require(
+            IERC20(erc20).balanceOf(to) >= balanceBefore.add(amount),
+            "Transfer amount is too low"
+        );
+    }
+
+    // ------------ PREDICTOOR ------------
+    function is_valid_subscription(address user) public view returns (bool) {
+        return subscriptions[user].expires <= block.number ? false : true;
+    }
+
+    function epoch(uint256 blocknum) public view returns (uint256) {
+        return blocknum / blocks_per_epoch;
+    }
+
+    function cur_epoch() public view returns (uint256) {
+        return epoch(block.number);
+    }
+
+    function rail_blocknum_to_slot(
+        uint256 blocknum
+    ) public view returns (uint256) {
+        uint256 rounded = blocknum / blocks_per_epoch;
+        return (rounded * blocks_per_epoch);
+    }   
+
+    function blocknum_is_on_a_slot(
+        uint256 blocknum
+    ) public view returns (bool) {
+        // a slot == beginning/end of an epoch
+        return blocknum == rail_blocknum_to_slot(blocknum);
+    }
+
+    function soonest_block_to_predict(uint256 prediction_block) public view returns (uint256) {
+        /*
+        Epoch i: predictoors submit predval for the beginning of epoch i+2. 
+        Predval is: "does trueval go UP or DOWN between the start of epoch i+1 and the start of epoch i+2?"
+        Once epoch i ends, predictoors cannot submit predvals for epoch i+2
+        */
+        return(rail_blocknum_to_slot(prediction_block)+ blocks_per_epoch * 2);
+    }
+
+    function submitted_predval(
+        uint256 blocknum,
+        address predictoor
+    ) public view returns (bool) {
+        uint256 slot = rail_blocknum_to_slot(blocknum);
+        return predobjs[slot][predictoor].predictoor != address(0);
+    }
+
+    function get_agg_predval(
+        uint256 blocknum
+    ) public view returns (uint256, uint256) {
+        require(is_valid_subscription(msg.sender), "No subscription");
+        uint256 slot = rail_blocknum_to_slot(blocknum);
+        return (agg_predvals_numer[slot], agg_predvals_denom[slot]);
+    }
+
+    function get_subscription_revenue_at_block(
+        uint256 blocknum
+    ) public view returns (uint256) {
+        uint256 slot = rail_blocknum_to_slot(blocknum);
+        return (subscription_revenue_at_block[slot]);
+    }
+
+    function get_prediction(
```

Comment:
Rename to `get_predval()`

Why: to disambiguate from each predictoor's `predval`, versus the top-level aggregated predval `agg_predval`.

Reply:
This returns entire prediction (predval,stake,predictoor,paid) and not just the predval